vnext/ReactUWP/Views/Image/ReactImage.cpp

```diff
@@ -65,51 +108,111 @@ winrt::fire_and_forget ReactImage::Source(ImageSource source) {
 
   winrt::Uri uri{Utf8ToUtf16(uriString)};
   winrt::hstring scheme{uri.SchemeName()};
-  bool needsDownload = (scheme == L"http") || (scheme == L"https");
+  bool needsDownload =
+      ((scheme == L"http") || (scheme == L"https")) && (m_useCompositionBrush || !source.headers.empty());
   bool inlineData = scheme == L"data";
+
   // get weak reference before any co_await calls
   auto weak_this{get_weak()};
 
   try {
     m_imageSource = source;
 
-    winrt::InMemoryRandomAccessStream memoryStream;
     if (needsDownload) {
-      memoryStream = co_await GetImageStreamAsync(source);
+      m_memoryStream = co_await GetImageStreamAsync(source);
     } else if (inlineData) {
-      memoryStream = co_await GetImageInlineDataAsync(source);
+      m_memoryStream = co_await GetImageInlineDataAsync(source);
     }
 
     if (auto strong_this{weak_this.get()}) {
-      if ((needsDownload || inlineData) && !memoryStream) {
+      if ((needsDownload || inlineData) && !strong_this->m_memoryStream) {
         strong_this->m_onLoadEndEvent(*strong_this, false);
       }
 
-      if (!needsDownload || memoryStream) {
-        auto surface = needsDownload || inlineData ? winrt::LoadedImageSurface::StartLoadFromStream(memoryStream)
-                                                   : winrt::LoadedImageSurface::StartLoadFromUri(uri);
-
-        strong_this->m_surfaceLoadedRevoker = surface.LoadCompleted(
-            winrt::auto_revoke,
-            [weak_this, surface](
-                winrt::LoadedImageSurface const & /*sender*/,
-                winrt::LoadedImageSourceLoadCompletedEventArgs const &args) {
-              if (auto strong_this{weak_this.get()}) {
-                bool succeeded{false};
-                if (args.Status() == winrt::LoadedImageSourceLoadStatus::Success) {
-                  strong_this->m_brush->Source(surface);
-                  succeeded = true;
-                }
-                strong_this->m_onLoadEndEvent(*strong_this, succeeded);
-              }
-            });
+      if (!needsDownload || strong_this->m_memoryStream) {
+        strong_this->SetBackground(needsDownload || inlineData, true);
       }
     }
   } catch (winrt::hresult_error const &) {
-    if (auto strong_this{weak_this.get()})
+    if (auto strong_this{weak_this.get()}) {
       strong_this->m_onLoadEndEvent(*strong_this, false);
+    }
   }
-} // namespace uwp
+}
+
+void ReactImage::SetBackground(bool fromStream, bool fireLoadEndEvent) {
+  const winrt::Uri uri{Utf8ToUtf16(m_imageSource.uri)};
+
+  if (m_useCompositionBrush) {
+    const auto compositionBrush = ReactImageBrush::Create();
+    compositionBrush->AvailableSize(m_availableSize);
+    compositionBrush->ResizeMode(m_resizeMode);
+
+    auto surface = fromStream ? winrt::LoadedImageSurface::StartLoadFromStream(m_memoryStream)
+                              : winrt::LoadedImageSurface::StartLoadFromUri(uri);
```

Comment:
At this point we may be waiting for m_memoryStream to get set, which will confuse SetBackground into thinking there's a valid URI.  I think SetBackground should be hardened to not try and load anything until after it has a valid stream or URI set.